server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java

```diff
@@ -216,7 +220,18 @@ protected void newResponseAsync(
 
     protected abstract NodeResponse newNodeResponse(StreamInput in, DiscoveryNode node) throws IOException;
 
-    protected abstract NodeResponse nodeOperation(NodeRequest request, Task task);
+    /**
+     * Implements the command logic on the node. The given `channel` may be used directly in the implementation, or the return object will
+     * be sent on the same channel: the implementation must do one or the other, never both.
+     *
+     * @param request
+     * @param channel The same transport channel on which returning a non-null response will be sent, but can alternatively be used directly
+     * to send a response (and then the implementation should return a null object).
+     * @param task
+     * @return Returns either a response to send on the TransportChannel, or null. If null, then the node logic handles the sending the
+     * response back on the given channel.
+     */
+    protected abstract @Nullable NodeResponse nodeOperation(NodeRequest request, TransportChannel channel, Task task);
```

Comment:
Hmm, the change here indeed causes a lot of disturbance and code changes and implications. And trying to think of alternatives.

## Alternative 1: make the async way the default

How about making the async way the default case for the base class?

I imagine e.g., making this `TransportNodesAction` have:

```
protected abstract void nodeAsyncOperation(NodeRequest request, ActionListener<NodeResponse> listener, Task task);
```

which will be exercised only by new commit notification at the moment (maybe more people will use it in the future).

And then making a new `TransportNodesSyncAction` which inherits `TransportNodesAction` and does:

```
protected abstract NodeResponse nodeOperation(NodeRequest request, Task task);

@Overrides
void nodeAsyncOperation(NodeRequest request, ActionListener<NodeResponse> listener, Task task) {
   ActionListener.completeWith(listener, () -> { // might need some thinking on which listener function here
      return nodeOperation(request, task);
   }
}
```

And so all other default classes would simply inherit TransportNodesSyncAction instead.

## Alternative 2: new async NodeTransportHandler

I see in the constructor that it does

```
transportService.registerRequestHandler(transportNodeAction, finalExecutor, nodeRequest, new NodeTransportHandler());
```

which uses the sync NodeTransportHandler. Maybe we could bypass this in the new commit notification and use our own node transport handler that would be async? That way we wouldn't need changes in any other classes.

I think I may prefer Alternative 1 though because it makes it possible to have both a sync and async flavor and people can choose.

Probably we can discuss more on this.

Reply:
Optionally passing the Handler in via the `TransportNodesAction` constructor -- say a `@Nullable` `Handler` argument -- crossed my mind. It felt like a blunt force hack, but appealing in the sense that it's quite obvious what's going on.

The idea of adding more inheritance -- e.g. `TransportNodesSyncAction` and `TransportNodesAsyncAction` -- to cover up deficiencies seems like more tech debt. It would make things less understandable. Though we could add documentation to help. 

We could take your `nodeAsyncOperation` method idea, have `nodeAsyncOperation` and `nodeSynchronousOperation`, and then add a `bool isAsync() { return false; }` method that child classes can override -- `doExecute()` would check the flag and call a method accordingly. This is appealing in the sense that child classes are up front about the choice, rather than it being an obscure implementation detail.

Atm, I'm leaning slightly towards a more feature-full base class than stretching logic across inheritance chains.

Reply:
I would be inclined to have an async method which by default just defers to the sync method that all existing implementations override:

```diff
diff --git a/server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java b/server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java
index 4fb24389170..f71c93eff0c 100644
--- a/server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java
+++ b/server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java
@@ -15,6 +15,7 @@ import org.elasticsearch.action.ActionListenerResponseHandler;
 import org.elasticsearch.action.FailedNodeException;
 import org.elasticsearch.action.support.ActionFilters;
 import org.elasticsearch.action.support.CancellableFanOut;
+import org.elasticsearch.action.support.ChannelActionListener;
 import org.elasticsearch.action.support.ThreadedActionListener;
 import org.elasticsearch.action.support.TransportAction;
 import org.elasticsearch.cluster.ClusterState;
@@ -218,6 +219,10 @@ public abstract class TransportNodesAction<

     protected abstract NodeResponse nodeOperation(NodeRequest request, Task task);

+    protected void nodeOperationAsync(NodeRequest request, Task task, ActionListener<NodeResponse> listener) {
+        ActionListener.completeWith(listener, () -> nodeOperation(request, task));
+    }
+
     /**
      * resolve node ids to concrete nodes of the incoming request
      **/
@@ -230,7 +235,7 @@ public abstract class TransportNodesAction<
     class NodeTransportHandler implements TransportRequestHandler<NodeRequest> {
         @Override
         public void messageReceived(NodeRequest request, TransportChannel channel, Task task) throws Exception {
-            channel.sendResponse(nodeOperation(request, task));
+            ActionListener.run(new ChannelActionListener<NodeResponse>(channel), l -> nodeOperationAsync(request, task, l));
         }
     }
```

This is similar to the `newResponse` and `newResponseAsync` methods we already have for this situation on the response side.